### ISSUE TITLE
Integrating Backbee into existing Symfony 2 project

### ISSUE THREAD

**Issue by gonzalovilaseca** [backbee/BackBee#590]
Hi,
We want to integrate Backee into our existing e-commerce platform based on Symfony2/Sylius.
We would replace all the CMF in place with Backbee.
What would be the best approach to have both Sylius and Backbee running in a same installation?
Thanks!

**Comment by mickaelandrieu** [comment]
To be honest,  I don't know :/ BackBee has not been developed as a bundle but as a complete application...  If you start something let us know and we will try to help. 

A first naive step could be to take a look to the EkinoWordpressBundle: share and make available data and BackBee services into Symfony application :)

**Comment by gonzalovilaseca** [comment]
Ok, will let you know thanks. Hopefully will start with it next week.

**Comment by gonzalovilaseca** [comment]
So I've started with this project and I have one question: Why a propietary `ControllerResolver` and not use `Symfony\Component\HttpKernel\Controller\ControllerResolver`.
(To integrate BB's routes I will have to hardcode the routes in a Symfony format as we can't have 2 different controller resolvers)

**Comment by mickaelandrieu** [comment]
ping @crouillon @eric-chau I guess because we need an instance of BBApplication in order to get the container...

**Comment by gonzalovilaseca** [comment]
Correct me if I'm wrong, but digging into the internals of BB backend it feels like you created a propietary framework using some of Symfony's components. 
Why not use Symfony framework straight away?

**Comment by ob7** [comment]
He probably packaged it for others hoping it would gain a momentum of its own.

**Comment by gonzalovilaseca** [comment]
I've started a full refactoring of BB, my goal is to make it work with Symfony framework.
I will try to separate what is infrastructure of what is not (DDD style).

**Comment by eric-chau** [comment]
Hello @gonzalovilaseca, I received your email but I decide to anwser you here, hope you don't mind. First, sorry for this late response. You're right, almost every Symfony component used are customized in BackBee. There are many reasons to this. The olders versions have been developed many years ago. We started the whole rework with Symfony2 1 years ago. It still in progress so obviously there are alot of aspect we can improve with a better integration of SF2 component.

**Comment by eric-chau** [comment]
Also, I can't wait to see your work about refactoring BackBee =)

**Comment by gonzalovilaseca** [comment]
Eric, thanks for the reply! 
So far I've been able to have Symfony detect all of BB doctrine annotations and create the full schema based on it. 
I've also created the following folder structure:
``BackBee\CoreDomain``: For core domain BB classes
``BackBee\CoreDomainBundle``: All `infrastructure` classes, so everything related to Symfony.

What I'm currently working on is:
1) Moving all of BB configuration loading/parsing etc to the Symfony way (using Configuration.php and *Extension.php clases).
2) Service definition and compilation in Symfony container.
3) Define listeners the Symfony way (as services).

**Comment by gonzalovilaseca** [comment]
@eric-chau If you don't mind I will ask you some questions via mail to help me speed up this refactoring :-)

**Comment by eric-chau** [comment]
This sounds good =)

**Comment by ConneXNL** [comment]
Any updates on this? I am kind of facing the problem where backbee would only be interesting if you can plug it into an existing SF2 project...

**Comment by mickaelandrieu** [comment]
@ConneXNL why? Drupal 8 is not a Symfony application or a bundle too. World don't have to be  Symfony framework centric: we have no interest to depend on the symfony/symfony metapackage. Instead of plug BackBee in a Symfony framework project, you can create real web applications in BackBee bundles like you will do in Symfony.

**Comment by gonzalovilaseca** [comment]
@ConneXNL Check this repo: https://github.com/ReissClothing/BackBee I've migrated BB to Symfony, but its still in pre alpha state. I'm currently working on integrating BB in our E-Commerce Symfony app, so will be updating the repo on a daily basis.

**Comment by gonzalovilaseca** [comment]
@mickaelandrieu Why not?

BB uses Symfony components to build a propietary Symfony like framework, I see no advantage in that, as the funtionality is the same in both frameworks.

By migrating to Symfony framework, you don't loose any of your current functionality, you get a framework that is tested, trusted and is constantly evolving, therefore you can focus on developing new BB features, not the framework itself. It also enforces you to follow best practices.

On top of that you open BB to a huge number of potential users, and that can translate into contributions that can improve BB in terms of features, fixes etc.

BB could not only be used as a standalone CMS, but as a content management bundle for any Symfony based app, and that opens a world of possibilities.

**Comment by mickaelandrieu** [comment]
@gonzalovilaseca it's not a *proprietary* framework, BackBee is under MIT licence. But above all, BackBee is a CMS, not a framework.

Symfony is a framework of course, but also  a list of **components**: it's not a bad practice to use it instead of always use a framework for an edge case.

It could be interesting to ask to the Drupal team why they choose to not use the Symfony framework.

**Comment by gonzalovilaseca** [comment]
@mickaelandrieu True, but BB is using almost all the Symfony components in a Symfony like framework, so why not use the framework itself, broadening BB's scope?

As I said, BB can be a full CMS solution with Symfony framework internals, that's the goal of my refactoring: allow BB to be used as a full CMS, but at the same time allow it to be ``plugged`` into any Symfony based project. 
At the same time I'm also separating BB core's classes from framework specific classes, so a migration to any other future framework will be possible.

Being on top of the Symfony framework devs will find a framework they're familiar with and that has plenty of documentation on the web.

I'm not trying to convince you :-), I'm just trying to explain my point of view.

**Comment by mickaelandrieu** [comment]
@gonzalovilaseca I'm already convinced :) tbh I wanted to move BackBee to Symfony Framework but we can't do that until 2.0 ^^^

We use 16 components of the 34, please don't depend on ``symfony/symfony`` metapackage, take a look Symfony empty edition: it's enough to be able to use all the bundles, and you don't need a metapackage to use the structure of a Symfony2|3 framework app :)

Good luck with your refactoring, asap I will try to help if I can.

**Comment by gonzalovilaseca** [comment]
@mickaelandrieu Ok, will do.
In the ``Issues`` section of the repo I have posted some RFC, and will keep adding new ones, so if you have time, please give your opinion.

**Comment by gonzalovilaseca** [comment]
@mickaelandrieu Also my intention is to use `Behat` and `phpspec` for tests, what do you think?

**Comment by mickaelandrieu** [comment]
@gonzalovilaseca I think theses tools are good, but I don't like Behat because I don't see the gain of Gherkin syntax except if your product owner write them.

**Comment by gonzalovilaseca** [comment]
@mickaelandrieu I find it useful even as a documentation of how things work. A dev getting into BB can get an idea of how things work by looking at the different scenarios, it's also a way of understanding what new PR's are doing.
I see Behat not only as a test tool, but also as a doc help.
